[ASTWalker] Continue post-walking with `Action::SkipChildren`

hamishknight opened the pull request:
Previously we would skip the post-walk, which was not a particularly intuitive behavior. Introduce an `Action::SkipNode` variant for walkers that want to also skip the post-walk. This then lets up clean up some walkers that previously had to work around the old behavior by doing the post-walk manually.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test

xedin commented on lib/Sema/CSGen.cpp:
```diff
@@ -848,11 +844,7 @@ namespace {
+
     /// Ignore statements.
     PreWalkResult<Stmt *> walkToStmtPre(Stmt *stmt) override {
       return Action::SkipChildren(stmt);
```
I think both Stmt and Decl walkers here should use `SkipChildrenAndPostWalk`.

xedin commented on lib/Sema/CSGen.cpp:
```diff
@@ -124,11 +124,7 @@ namespace {
-      return Action::Continue(expr);
-    }
-    
+
```
All of the SkipChildren here should be `SkipChildrenAndPostWalk`, I don't think we care about post walk for any of the ignored things...

hamishknight commented:
After talking with Pavel, I'm considering renaming `Action::SkipChildrenAndPostWalk` to something like `Action::SkipNode`, and making it the norm for ASTWalkers that don't care about the post-walk

bnbarham reviewed (approved):
Very nice 😍 

> I'm considering renaming Action::SkipChildrenAndPostWalk to something like Action::SkipNode, and making it the norm for ASTWalkers that don't care about the post-walk

This sounds nice to me.

hamishknight commented:
> After talking with Pavel, I'm considering renaming Action::SkipChildrenAndPostWalk to something like Action::SkipNode, and making it the norm for ASTWalkers that don't care about the post-walk

Gone ahead and made this change

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

xedin commented:
Looks great to me, thank you!

hamishknight commented:
Going to hold off merging this for a little bit until we've stopped cherry-picking stuff to 5.10

hamishknight committed:
NFC: Rename TypeWalker's `SkipChildren` to `SkipNode`

For consistency with ASTWalker.

hamishknight committed:
[ASTWalker] NFC: Rename `SkipChildren` -> `SkipNode`

This better describes what the action currently
does, and allows us to re-introduce `SkipChildren`
with the correct behavior.

hamishknight committed:
[ASTWalker] Re-introduce `Action::SkipChildren`

Now with the behavior of resuming at the post-walk.

hamishknight committed:
Remove some redundant post walks

hamishknight committed:
Remove PreWalkAction/PostWalkAction constructors

Help avoid using e.g `PreWalkAction::SkipChildren`
directly, and reduce duplication of walk actions to
take.

hamishknight committed:
[ASTWalker] Cleanup some cases manually doing post-walks

We can replace these with `Action::SkipChildren`.

hamishknight committed:
[IDE] Add some FIXMEs

hamishknight commented:
@swift-ci please test

hamishknight commented:
🚢
